db.go

```diff
@@ -0,0 +1,225 @@
+//go:build exclude
+
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	"math"
+	"os"
+	"strings"
+	"time"
+	"unicode"
+
+	_ "github.com/mattn/go-sqlite3"
+)
+
+type saveDb struct {
+	db        *sql.DB
+	dbPath    string
+	tableName string
+}
+
+const (
+	eventTypeStatistics     = "statistics"
+	eventTypeHostnameChange = "hostname change"
+	tableSchema             = `
+-- Organized row names together for better readability
+CREATE TABLE %s (
+    id INTEGER PRIMARY KEY,
+    event_type TEXT NOT NULL, -- for the data type eg. statistics, hostname change
+    timestamp DATETIME,
+    addr TEXT,
+    hostname TEXT,
+    port INTEGER,
+    hostname_resolve_tries INTEGER,
+
+    hostname_changed_to TEXT,
+    hostname_change_time DATETIME,
+
+    latency_min REAL,
+    latency_avg REAL,
+    latency_max REAL,
+
+	total_duration TEXT,
+    start_timestamp DATETIME,
+    end_timestamp DATETIME,
+
+	never_succeed_probe INTEGER, -- value will be 1 if a prove never succeeded
+	never_failed_probe INTEGER, -- value will be 1 if a prove never failed
+    last_successful_probe DATETIME,
+    last_unsuccessful_probe DATETIME,
+
+    longest_uptime TEXT,
+    longest_uptime_end DATETIME,
+    longest_uptime_start DATETIME,
+
+    longest_downtime TEXT,
+    longest_downtime_start DATETIME,
+    longest_downtime_end DATETIME,
+
+    total_packets INTEGER,
+    total_packet_loss REAL,
+    total_successful_probes INTEGER,
+    total_unsuccessful_probes INTEGER,
+
+    total_uptime TEXT,
+    total_downtime TEXT
+);`
+)
+
+// newDb creas a newDb to the given path and returns `saveDb` struct
+func newDb(args []string, dbPath string) saveDb {
+	tableName := newTableName(args)
+	tableSchema := fmt.Sprintf(tableSchema, tableName)
+
+	db, err := sql.Open("sqlite3", dbPath)
+	if err != nil {
+		colorRed("\nwhile creating the database %q: %s\n", dbPath, err)
+		os.Exit(1)
+	}
+
+	_, err = db.Exec(tableSchema)
+	if err != nil {
+		colorRed("\nwhile writing to the detabase %q \nerr: %s\n", dbPath, err)
+		os.Exit(1)
+	}
+
+	return saveDb{db, dbPath, tableName}
+}
+
+// newTableName will return correctly formatted table name
+func newTableName(args []string) string {
+	// table name can't have '.'
+	// formating the table name "example_com_hour_minute_sec_day_month_year"
+	tableName := fmt.Sprintf("%s_%s_%s", strings.ReplaceAll(args[0], ".", "_"), args[1], time.Now().Format("15_04_05_01_02_2006"))
+
+	// table name can't start with numbers
+	if unicode.IsNumber(rune(tableName[0])) {
+		tableName = "_" + tableName
+	}
+
+	return tableName
+}
+
+// this will insert the table name and
+// save the ags to the database
+func (s saveDb) save(query string, args ...any) error {
+	// inserting the table name
+	statement := fmt.Sprintf(query, s.tableName)
+
+	// saving to the db
+	_, err := s.db.Exec(statement, args...)
+
```

Comment:
We never close the DB connection. This leads to resource leak and is unsafe overall.